src/Builder/BuilderEncoder.php

```diff
@@ -4,323 +4,101 @@
-        $result = new stdClass();
-        $result->_id = $this->recursiveEncode($value->_id);
-
-        foreach (get_object_vars($value->field) as $key => $val) {
-            $result->{$key} = $this->recursiveEncode($val);
-        }
-
-        return $this->wrap($value, $result);
-    }
-
-    private function encodeAsObject(OperatorInterface $value): stdClass
-    {
-        $result = new stdClass();
-        foreach (get_object_vars($value) as $key => $val) {
-            // Skip optional arguments. If they have a default value, it is resolved by the server.
-            if ($val === Optional::Undefined) {
-                continue;
-            }
-
-            $result->{$key} = $this->recursiveEncode($val);
-        }
-
-        return $this->wrap($value, $result);
-    }
-
-    private function encodeAsDollarObject(OperatorInterface $value): stdClass
-    {
-        $result = new stdClass();
-        foreach (get_object_vars($value) as $key => $val) {
-            // Skip optional arguments. If they have a default value, it is resolved by the server.
-            if ($val === Optional::Undefined) {
-                continue;
-            }
-
-            $val = $this->recursiveEncode($val);
-
-            if ($key === 'geometry') {
-                if (is_object($val) && property_exists($val, '$geometry')) {
-                    $result->{'$geometry'} = $val->{'$geometry'};
-                } elseif (is_array($val) && array_key_exists('$geometry', $val)) {
-                    $result->{'$geometry'} = $val->{'$geometry'};
-                } else {
-                    $result->{'$geometry'} = $val;
-                }
-            } else {
-                $result->{'$' . $key} = $val;
-            }
+        if (! is_object($value)) {
+            return false;
         }
 
-        return $this->wrap($value, $result);
-    }
-
-    /**
-     * Get the unique property of the operator as value
-     */
-    private function encodeAsSingle(OperatorInterface $value): stdClass
-    {
-        foreach (get_object_vars($value) as $val) {
-            $result = $this->recursiveEncode($val);
-
-            return $this->wrap($value, $result);
-        }
-
-        throw new LogicException(sprintf('Class "%s" does not have a single property.', $value::class));
-    }
-
-    private function encodeCombinedFilter(CombinedFieldQuery $filter): stdClass
-    {
-        $result = new stdClass();
-        foreach ($filter->fieldQueries as $filter) {
-            $filter = $this->recursiveEncode($filter);
-            if (is_object($filter)) {
-                $filter = get_object_vars($filter);
-            } elseif (! is_array($filter)) {
-                throw new LogicException(sprintf('Query filters must an array or an object. Got "%s"', get_debug_type($filter)));
-            }
-
-            foreach ($filter as $key => $value) {
-                $result->{$key} = $value;
-            }
-        }
+        $encoder = $this->getEncoderFor($value);
```

Comment:
This reminds me of the original implementation of codec libraries (https://github.com/mongodb/mongo-php-library/pull/1160), which were removed from PHPLIB before release.

Reply:
Yes. A primary concern in that implementation was iterating the list of encoders and calling `canEncode` on each of them. Here we have a list of known encoders and the classes they should be able to encode, so we can decide on a codec upfront.